core-schema/src/schema.rs

```diff
@@ -0,0 +1,585 @@
+
+    /// validation errors
+    pub errors: Vec<SchemaError>,
+}
+
+impl<'a> Schema<'a> {
+    /// Parse a schema source string and return a validated `Schema` with all
+    /// `Request`s identified, or fail with a `graphql_parser::ParseError`.
+    ///
+    /// As indicated by the type, this fn will always return `Ok` if parsing
+    /// succeeds. Subsequent errors are reported in the `errors` field of the
+    /// returned schema.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use core_schema::{Schema, Feature, Spec, Version};
+    /// use graphql_parser::{ParseError, Pos};
+    /// use std::borrow::Cow;
+    ///
+    /// let schema = Schema::parse(r#"
+    ///     schema
+    ///       @core(feature: "https://lib.apollo.dev/core/v0.1")
+    ///       @core(feature: "https://spec.example.com/A/v1.2", as: "exampleA")
+    ///       @core(feature: "https://example.net/specB/v0.1")
+    ///     {
+    ///       query: Query
+    ///     }
+    /// "#)?;
+    ///
+    /// assert_eq!(schema.errors, vec![]);
+    /// assert_eq!(schema.features, vec![
+    ///    Feature {
+    ///        spec: Spec::new("https://lib.apollo.dev/core", "core", (0, 1)),
+    ///        name: Cow::Borrowed("core"),
+    ///        position: Pos { line: 3, column: 7 },
+    ///    },        
+    ///    Feature {
+    ///        spec: Spec::new("https://spec.example.com/A", "A", (1, 2)),
+    ///        name: Cow::Borrowed("exampleA"),
+    ///        position: Pos { line: 4, column: 7 },
+    ///    },
+    ///    Feature {
+    ///        spec: Spec::new("https://example.net/specB", "specB", (0, 1)),
+    ///        name: Cow::Borrowed("specB"),
+    ///        position: Pos { line: 5, column: 7 },
+    ///    },
+    /// ]);
+    /// # Ok::<(), ParseError>(())
+    /// ```
+    pub fn parse(source: &'a str) -> Result<Schema<'a>, ParseError> {
+        let document = parse_schema(source)?;
+        let mut errors = Vec::new();
+
+        let mut schemas: Vec<_> = document
+            .definitions
+            .iter()
+            .filter_map(|def| match def {
+                Definition::Schema(sd) => Some(sd),
+                _ => None,
+            })
+            .collect();
+
+        if schemas.is_empty() {
+            errors.push(SchemaError::NoSchemas);
+
+            // We only return `Err` for actual parse errors. Validation failures
+            // such as NoSchemas are available on the `errors` field.
+            return Ok(Schema {
+                document,
+                features: vec![],
+                errors,
+            });
+        }
+        let schema = schemas.remove(0);
+        for extraneous in schemas {
+            errors.push(SchemaError::ExtraSchema(extraneous.position));
+        }
+
+        // Collect everything which *could be* a core feature declaration.
+        // "Could be a core feature" means it's a directive on
+        // the schema which has a `spec: String` argument. We consider
```

Comment:
s/spec/feature/

and that's more like "a spec argument whose value is a String" (vs wondering if the argument could be `String!`)